### PR TITLE
Add binary code challenge 

### LINKED ISSUE #39: Add Binary Challenge
Given two binary strings a and b, return their sum as a binary string.

Example 1:

Input: a = "11", b = "1"
Output: "100"
Example 2:

Input: a = "1010", b = "1011"
Output: "10101"
 

Constraints:

1 <= a.length, b.length <= 104
a and b consist only of '0' or '1' characters.
Each string does not contain leading zeros except for the zero itself.

### PR DESCRIPTION
**Pull request by SiSaR-Pal**
---
name: solution review 
about: A template PR for code review with a checklist
---
* Solving #39 

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/add_binary.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    - Input strings must not exceed 10^4 characters in length.
+
+Assumptions:
+    - Leading zeros in the input do not affect the result.
+    - Input strings are well-formed binary strings without embedded whitespace.
+
+Efficiency:
+    - The function performs efficiently for binary strings up to 10^4 characters,
+      leveraging Python's built-in `int` and `bin` for conversion and calculation.
+"""
```
Comment (Nada-saad635) [inline solutions/add_binary.py:22]: as a beginner  i feel a little bit difficult to understand what function do , can you add more explanation ??

**Comment by Nada-saad635** [comment]
ANOTHER THING  There is no  a module docstring in the tests file

**Comment by Nada-saad635** [comment]
thats it ! very hard work , its really good ! good job

**Comment by Elshikh-Amro** [comment]
great work man it is well written 
but it is a bit complicated function that requires some step-by-step explanation if you can do that

File: solutions/add_binary.py (hunk 1 above)
Comment (SiSaR-Pal) [reply solutions/add_binary.py:22]: I will try to simplify the documentation

**Comment by SiSaR-Pal** [comment]
I have made the requested modifications to the add-binary-code branch based on your feedback.
Please let me know if there are any additional concerns or adjustments needed.

File: solutions/tests/test_add_binary.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+"""
```
Comment (heba3) [inline solutions/tests/test_add_binary.py:1]: The function is well-designed, efficiently handles binary addition, and includes robust input validation. Its clear documentation and example doctests enhance usability. 
I recommend splitting the tests with multiple assertions into individual tests for clarity and easier debugging.

File: solutions/tests/test_add_binary.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        """Test addition of binary strings with leading zeros."""
+        self.assertEqual(add_binary("00011", "1"), "100")
+
+    def test_large_binary_strings(self):
+        """Test addition of very large binary strings."""
+        bin_num1 = "1" * 10000
+        bin_num2 = "1"
+        expected = "1" + "0" * 10000
+        self.assertEqual(add_binary(bin_num1, bin_num2), expected)
+
```
Comment (heba3) [inline solutions/tests/test_add_binary.py:60]:
``` python 
def test_empty_input_first(self):
    """Test that a ValueError is raised when the first input is empty."""
    with self.assertRaises(ValueError):
        add_binary("", "1")

def test_empty_input_second(self):
    """Test that a ValueError is raised when the second input is empty."""
    with self.assertRaises(ValueError):
        add_binary("1", "")

File: solutions/tests/test_add_binary.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    def test_empty_input(self):
+        """Test that a ValueError is raised for empty input strings."""
+        with self.assertRaises(ValueError):
+            add_binary("", "1")
+        with self.assertRaises(ValueError):
+            add_binary("1", "")
+
```
Comment (heba3) [inline solutions/tests/test_add_binary.py]: Same here, 
```python 
def test_invalid_characters_first(self):
    """Test that a ValueError is raised when the first input contains invalid characters."""
    with self.assertRaises(ValueError):
        add_binary("102", "1")

def test_invalid_characters_second(self):
    """Test that a ValueError is raised when the second input contains invalid characters."""
    with self.assertRaises(ValueError):
        add_binary("11a", "1")

File: solutions/tests/test_add_binary.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    def test_invalid_characters(self):
+        """Test that a ValueError is raised for inputs with non-binary characters."""
+        with self.assertRaises(ValueError):
+            add_binary("102", "1")
+        with self.assertRaises(ValueError):
+            add_binary("11a", "1")
+
```
Comment (heba3) [inline solutions/tests/test_add_binary.py]:
```python
def test_whitespace_in_input_first(self):
    """Test that a ValueError is raised when the first input contains whitespace."""
    with self.assertRaises(ValueError):
        add_binary(" 1010 ", "101")

def test_whitespace_in_input_second(self):
    """Test that a ValueError is raised when the second input contains whitespace."""
    with self.assertRaises(ValueError):
        add_binary("1010", " 101 ")

**Review by heba3 — COMMENTED** [review]
Good job Said! and nice work.

**Review by Elshikh-Amro — COMMENTED** [review]
it is well written and formatted gooood job!!!!!!
my only take is  assertions you can divide them into separate tests

**Comment by SiSaR-Pal** [comment]
Thanks a lot for your thorough review and thoughtful remarks

**Review by heba3 — APPROVED** [review]
